Update to mod adds dependency, CKAN upgrade doesn't install it?

I apologise if this issue is already open, I didn't see it when I searched. I'm using v1.6.23-0. It looks like Outer Planets Mod has somehow changed its dependency on Kopernicus (possibly due to Kopernicus 0.9 pre-release?). OPM is listed as updatable in the gui, so I tell it to upgrade and the progress bar just spins forever and it never finishes. I attempted to upgrade on the command line using ckan upgrade --all, and it downloaded the mod but I get a stack trace when it tries to install:

```
Unhandled Exception:
The following inconsistencies were found:
OuterPlanetsMod requires Kopernicus but nothing provides it  at CKAN.SanityChecker.EnforceConsistency (IEnumerable`1 modules, IEnumerable`1 dlls) [0x00000] in <filename unknown>:0 
  at CKAN.Registry.CheckSanity () [0x00000] in <filename unknown>:0 
  at CKAN.RegistryManager.Save (Boolean enforce_consistency) [0x00000] in <filename unknown>:0 
  at CKAN.ModuleInstaller.AddRemove (IEnumerable`1 add, IEnumerable`1 remove) [0x00000] in <filename unknown>:0 
  at CKAN.ModuleInstaller.Upgrade (IEnumerable`1 modules, CKAN.NetAsyncDownloader netAsyncDownloader) [0x00000] in <filename unknown>:0 
  at CKAN.CmdLine.Upgrade.RunCommand (CKAN.KSP ksp, System.Object raw_options) [0x00000] in <filename unknown>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: The following inconsistencies were found:
OuterPlanetsMod requires Kopernicus but nothing provides it  at CKAN.SanityChecker.EnforceConsistency (IEnumerable`1 modules, IEnumerable`1 dlls) [0x00000] in <filename unknown>:0 
  at CKAN.Registry.CheckSanity () [0x00000] in <filename unknown>:0 
  at CKAN.RegistryManager.Save (Boolean enforce_consistency) [0x00000] in <filename unknown>:0 
  at CKAN.ModuleInstaller.AddRemove (IEnumerable`1 add, IEnumerable`1 remove) [0x00000] in <filename unknown>:0 
  at CKAN.ModuleInstaller.Upgrade (IEnumerable`1 modules, CKAN.NetAsyncDownloader netAsyncDownloader) [0x00000] in <filename unknown>:0 
  at CKAN.CmdLine.Upgrade.RunCommand (CKAN.KSP ksp, System.Object raw_options) [0x00000] in <filename unknown>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

A few humble opinions:
1. It should not be possible for the gui to spin forever with no notification as to what's going on. An exception is thrown, so the problem is definitely detectable, so the user should be notified that a problem occurred and the process be cancelled. This should apply regardless of what caused the exception, as this is the second separate issue I've encountered that resulted in this behavior.
2. Kopernicus and KopernicusTech are both installable, but neither of them are currently installed. I infer from this that the dependency on Kopernicus et al was a recent addition to the OPM metadata. CKAN should detect that a new dependency exists when trying to upgrade and add the new dependency to the list of mods to be installed, rather than attempting to upgrade OPM and find out half way through that there is an unmet dependency.
3. Using CKAN is still way better than trying to manage the dependencies by hand. Thank you all for all your hard work.

--- a/Cmdline/Action/Upgrade.cs
+++ b/Cmdline/Action/Upgrade.cs
@@ -41,7 +41,7 @@ namespace CKAN.CmdLine
                 options.modules.Add(MainClass.LoadCkanFromFile(ksp, options.ckan_file).identifier);
             }
 
-            if (options.modules.Count == 0 && ! options.upgrade_all)
+            if (options.modules.Count == 0 && !options.upgrade_all)
             {
                 // What? No files specified?
                 User.RaiseMessage("Usage: ckan upgrade Mod [Mod2, ...]");
@@ -90,57 +90,50 @@ namespace CKAN.CmdLine
                 var registry = regMgr.registry;
                 if (options.upgrade_all)
                 {
-                    var installed = new Dictionary<string, ModuleVersion>(registry.Installed());
                     var to_upgrade = new List<CkanModule>();
 
-                    foreach (KeyValuePair<string, ModuleVersion> mod in installed)
+                    foreach (KeyValuePair<string, ModuleVersion> mod in registry.Installed(false))
                     {
-                        ModuleVersion current_version = mod.Value;
-
-                        if ((current_version is ProvidesModuleVersion) || (current_version is UnmanagedModuleVersion))
+                        try
                         {
-                            continue;
+                            // Check if upgrades are available
+                            var latest = registry.LatestAvailable(mod.Key, ksp.VersionCriteria());
+
+                            // This may be an unindexed mod. If so,
+                            // skip rather than crash. See KSP-CKAN/CKAN#841.
+                            if (latest == null || latest.IsDLC)
+                            {
+                                continue;
+                            }
+
+                            if (latest.version.IsGreaterThan(mod.Value))
+                            {
+                                // Upgradable
+                                log.InfoFormat("New version {0} found for {1}",
+                                    latest.version, latest.identifier);
+                                to_upgrade.Add(latest);
+                            }
+
                         }
-                        else
+                        catch (ModuleNotFoundKraken)
                         {
-                            try
-                            {
-                                // Check if upgrades are available
-                                var latest = registry.LatestAvailable(mod.Key, ksp.VersionCriteria());
-
-                                // This may be an unindexed mod. If so,
-                                // skip rather than crash. See KSP-CKAN/CKAN#841.
-                                if (latest == null || latest.IsDLC)
-                                {
-                                    continue;
-                                }
-
-                                if (latest.version.IsGreaterThan(mod.Value))
-                                {
-                                    // Upgradable
-                                    log.InfoFormat("New version {0} found for {1}",
-                                        latest.version, latest.identifier);
-                                    to_upgrade.Add(latest);
-                                }
-
-                            }
-                            catch (ModuleNotFoundKraken)
-                            {
-                                log.InfoFormat("{0} is installed, but no longer in the registry",
-                                    mod.Key);
-                            }
+                            log.InfoFormat("{0} is installed, but no longer in the registry",
+                                mod.Key);
                         }
-
                     }
 
-                    ModuleInstaller.GetInstance(ksp, manager.Cache, User).Upgrade(to_upgrade, new NetAsyncModulesDownloader(User, manager.Cache), ref possibleConfigOnlyDirs, regMgr);
+                    ModuleInstaller.GetInstance(ksp, manager.Cache, User).Upgrade(to_upgrade, new NetAsyncModulesDownloader(User, manager.Cache), ref possibleConfigOnlyDirs, regMgr, true, true);
                 }
                 else
                 {
-                    // TODO: These instances all need to go.
                     Search.AdjustModulesCase(ksp, options.modules);
                     ModuleInstaller.GetInstance(ksp, manager.Cache, User).Upgrade(options.modules, new NetAsyncModulesDownloader(User, manager.Cache), ref possibleConfigOnlyDirs, regMgr);
                 }
+            }
+            catch (CancelledActionKraken k)
+            {
+                User.RaiseMessage("Upgrade aborted: {0}", k.Message);
+                return Exit.ERROR;
             }
             catch (ModuleNotFoundKraken kraken)
             {

--- a/Cmdline/ConsoleUser.cs
+++ b/Cmdline/ConsoleUser.cs
@@ -41,7 +41,7 @@ namespace CKAN.CmdLine
                 return true;
             }
 
-            Console.Write("{0} [Y/n] ", question);
+            Console.Write("\r\n{0} [Y/n] ", question);
             while (true)
             {
                 var input = Console.In.ReadLine();

--- a/ConsoleUI/ModListScreen.cs
+++ b/ConsoleUI/ModListScreen.cs
@@ -272,9 +272,6 @@ namespace CKAN.ConsoleUI {
                 new ConsoleMenuOption("Audit recommendations",      "",
                     "List mods suggested and recommended by installed mods",
                     true, ViewSuggestions),
-                new ConsoleMenuOption("Scan KSP dir",               "",
-                    "Check for manually installed mods",
-                    true, ScanForMods),
                 new ConsoleMenuOption("Import downloads...",        "",
                     "Select manually downloaded mods to import into CKAN",
                     true, ImportDownloads),
@@ -352,8 +349,8 @@ namespace CKAN.ConsoleUI {
 
         private bool HasAnyUpgradeable()
         {
-            foreach (InstalledModule im in registry.InstalledModules) {
-                if (registry.HasUpdate(im.identifier, manager.CurrentInstance.VersionCriteria())) {
+            foreach (string identifier in registry.Installed(true).Select(kvp => kvp.Key)) {
+                if (registry.HasUpdate(identifier, manager.CurrentInstance.VersionCriteria())) {
                     return true;
                 }
             }
@@ -362,9 +359,9 @@ namespace CKAN.ConsoleUI {
 
         private bool UpgradeAll()
         {
-            foreach (InstalledModule im in registry.InstalledModules) {
-                if (registry.HasUpdate(im.identifier, manager.CurrentInstance.VersionCriteria())) {
-                    plan.Upgrade.Add(im.identifier);
+            foreach (string identifier in registry.Installed(true).Select(kvp => kvp.Key)) {
+                if (registry.HasUpdate(identifier, manager.CurrentInstance.VersionCriteria())) {
+                    plan.Upgrade.Add(identifier);
                 }
             }
             return true;
@@ -507,6 +504,7 @@ namespace CKAN.ConsoleUI {
 
         private List<CkanModule> GetAllMods(bool force = false)
         {
+            ScanForMods();
             if (allMods == null || force) {
                 allMods = new List<CkanModule>(registry.CompatibleModules(manager.CurrentInstance.VersionCriteria()));
                 foreach (InstalledModule im in registry.InstalledModules) {
@@ -716,9 +714,7 @@ namespace CKAN.ConsoleUI {
         public InstallStatus GetModStatus(KSPManager manager, IRegistryQuerier registry, string identifier)
         {
             if (registry.IsInstalled(identifier, false)) {
-                if (registry.IsAutodetected(identifier)) {
-                    return InstallStatus.AutoDetected;
-                } else if (Remove.Contains(identifier)) {
+                if (Remove.Contains(identifier)) {
                     return InstallStatus.Removing;
                 } else if (registry.HasUpdate(identifier, manager.CurrentInstance.VersionCriteria())) {
                     if (Upgrade.Contains(identifier)) {
@@ -726,6 +722,8 @@ namespace CKAN.ConsoleUI {
                     } else {
                         return InstallStatus.Upgradeable;
                     }
+                } else if (registry.IsAutodetected(identifier)) {
+                    return InstallStatus.AutoDetected;
                 } else if (Replace.Contains(identifier)) {
                     return InstallStatus.Replacing;
                 } else if (registry.GetReplacement(identifier, manager.CurrentInstance.VersionCriteria()) != null) {

--- a/Core/KSP.cs
+++ b/Core/KSP.cs
@@ -476,7 +476,9 @@ namespace CKAN
                 bool dlcChanged = manager.ScanDlc();
 
                 if (dllChanged || dlcChanged)
+                {
                     manager.Save(false);
+                }
 
                 tx.Complete();
 

--- a/Core/ModuleInstaller.cs
+++ b/Core/ModuleInstaller.cs
@@ -1004,12 +1004,6 @@ namespace CKAN
 
                 if (installed_mod == null)
                 {
-                    //Maybe ModuleNotInstalled ?
-                    if (registry_manager.registry.IsAutodetected(ident))
-                    {
-                        throw new ModuleNotFoundKraken(ident, module.version.ToString(), String.Format("Can't upgrade {0} as it was not installed by CKAN. \r\n Please remove manually before trying to install it.", ident));
-                    }
-
                     User.RaiseMessage("Installing previously uninstalled mod {0}", ident);
                 }
                 else

--- a/Core/Registry/IRegistryQuerier.cs
+++ b/Core/Registry/IRegistryQuerier.cs
@@ -184,7 +184,6 @@ namespace CKAN
             }
             if (newest_version == null
                 || !querier.IsInstalled(identifier, false)
-                || querier.InstalledDlls.Contains(identifier)
                 || !newest_version.version.IsGreaterThan(querier.InstalledVersion(identifier)))
             {
                 return false;

--- a/GUI/Main/Main.cs
+++ b/GUI/Main/Main.cs
@@ -372,6 +372,7 @@ namespace CKAN
         /// <param name="allowRepoUpdate">true if a repo update is allowed if needed (e.g. on initial load), false otherwise</param>
         private void CurrentInstanceUpdated(bool allowRepoUpdate)
         {
+            CurrentInstance.ScanGameData();
             Util.Invoke(this, () =>
             {
                 Text = $"CKAN {Meta.GetVersion()} - KSP {CurrentInstance.Version()}    --    {CurrentInstance.GameDir().Replace('/', Path.DirectorySeparatorChar)}";

--- a/GUI/Model/GUIMod.cs
+++ b/GUI/Model/GUIMod.cs
@@ -111,9 +111,6 @@ namespace CKAN
         /// </returns>
         public bool IsInstallable()
         {
-            // Auto detected mods are never installable
-            if (IsAutodetected)
-                return false;
             // Compatible mods are installable, but so are mods that are already installed
             return !IsIncompatible || IsInstalled;
         }
@@ -203,7 +200,7 @@ namespace CKAN
             Identifier     = identifier;
             IsAutodetected = registry.IsAutodetected(identifier);
             DownloadCount  = registry.DownloadCount(identifier);
-            if (registry.IsAutodetected(identifier))
+            if (IsAutodetected)
             {
                 IsInstalled = true;
             }

--- a/GUI/Model/ModList.cs
+++ b/GUI/Model/ModList.cs
@@ -265,7 +265,12 @@ namespace CKAN
 
             ModChange myChange = changes?.FindLast((ModChange ch) => ch.Mod.Equals(mod));
 
-            var selecting = mod.IsInstallable()
+            var selecting = mod.IsAutodetected
+                ? new DataGridViewTextBoxCell()
+                {
+                    Value = Properties.Resources.MainModListAutoDetected
+                }
+                : mod.IsInstallable()
                 ? (DataGridViewCell) new DataGridViewCheckBoxCell()
                 {
                     Value = myChange == null ? mod.IsInstalled
@@ -275,7 +280,7 @@ namespace CKAN
                 }
                 : new DataGridViewTextBoxCell()
                 {
-                    Value = mod.IsAutodetected ? Properties.Resources.MainModListAutoDetected : "-"
+                    Value = "-"
                 };
 
             var autoInstalled = mod.IsInstalled && !mod.IsAutodetected

